### PR TITLE
MSO DHCP Policies

### PR DESCRIPTION
**Pull request by jorgegomez2307**
Added support to create/delete/modify DHCP Relay Policies and DHCP Relay Policies Providers
State 'Query' still to be discussed

Please reach out to me internally for further questions (CCO: jgomezve)

### CODE REVIEW

File: plugins/modules/mso_dhcp_relay_policy.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+# Copyright: (c) 2020, Jorge Gomez Velasquez <jgomezve@cisco.com>
+
+ANSIBLE_METADATA = {"metadata_version": "0.1", "status": ["preview"]}
+
+DOCUMENTATION = r"""
+---
+module: mso_dhcp_relay_policy
+short_description: Configure a DHCP Relay policy.
+description:
+- Configure a DHCP Relay policy.
```
Comment (shrsr) [inline plugins/modules/mso_dhcp_relay_policy.py:21]: Is version required?

File: plugins/modules/mso_dhcp_relay_policy.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+version_added: '2.10'
+options:
```
Comment (shrsr) [inline plugins/modules/mso_dhcp_relay_policy.py:24]: Is version required?

File: plugins/modules/mso_dhcp_relay_policy.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  state:
+    description:
+    - State of the DHCP Relay Policy
+    type: str
+    required: yes
+author:
+- Jorge Gomez Velasquez
+"""
+
+EXAMPLES = r"""
```
Comment (shrsr) [inline plugins/modules/mso_dhcp_relay_policy.py:46]: - name: Add a new DHCP Relay Policy 1 (check mode)
   cisco.mso.mso_dhcp_relay_policy:
    name: DHCP_REL_POL_1
    description: "My Test DHCP Policy 1"
    tenant: ansible_test
    state: present
  delegate_to: localhost

- name: Query  a DHCP Relay Policy 2 
   cisco.mso.mso_dhcp_relay_policy:
    name: DHCP_REL_POL_2
    state: query
   delegate_to: localhost

- name: Query  all
   cisco.mso.mso_dhcp_relay_policy:
    state: query
  delegate_to: localhost

- name: Remove DHCP Relay Policy 1  (normal mode)
   cisco.mso.mso_dhcp_relay_policy:
    name: DHCP_REL_POL_1
    description: "My Test DHCP Policy 1"
    tenant: ansible_test
    state: absent
  delegate_to: localhost

File: plugins/modules/mso_dhcp_relay_policy.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        tenantId=tenant_id
+    )
+
+    response = {}
+    changed = False
+
+    if state == 'absent':
+        if mso.existing:
+            mso.existing = mso.request(path, method='DELETE', data=mso.sent)
+            changed = True
```
Comment (shrsr) [inline plugins/modules/mso_dhcp_relay_policy.py:117]: mso.py takes care of this

File: plugins/modules/mso_dhcp_relay_policy.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+        if mso.existing:
+            if mso.check_changed():
+                mso.existing = mso.request(path, method='PUT', data=mso.sent)
+                changed = True
+                response['name'] = mso.existing['name']
+                response['tenant'] = mso.existing['tenantId']
+        else:
+            mso.existing = mso.request(path, method='POST', data=mso.sent)
+            changed = True
```
Comment (shrsr) [inline plugins/modules/mso_dhcp_relay_policy.py:131]: mso.py takes care of this

File: plugins/modules/mso_dhcp_relay_policy.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    payload = dict(
+        name=name,
+        desc=description,
+        policyType="dhcp",
+        policySubtype="relay",
+        tenantId=tenant_id
+    )
+
+    response = {}
+    changed = False
```
Comment (shrsr) [inline plugins/modules/mso_dhcp_relay_policy.py:109]: Is this required?

File: plugins/modules/mso_dhcp_relay_policy.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+    if state == 'absent':
+        if mso.existing:
+            mso.existing = mso.request(path, method='DELETE', data=mso.sent)
```
Comment (shrsr) [inline plugins/modules/mso_dhcp_relay_policy.py]: Can be changed from  mso.existing = mso.request(path, method='DELETE', data=mso.sent)
to  mso.request(path, method='DELETE', data=mso.sent)

mso.py does the rest

File: plugins/modules/mso_dhcp_relay_policy.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            changed = True
+    elif state == 'present':
+
+        mso.sanitize(payload, collate=True)
+
+        if mso.existing:
+            if mso.check_changed():
+                mso.existing = mso.request(path, method='PUT', data=mso.sent)
```
Comment (shrsr) [inline plugins/modules/mso_dhcp_relay_policy.py]: Same as above

File: plugins/modules/mso_dhcp_relay_policy.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+                changed = True
+                response['name'] = mso.existing['name']
+                response['tenant'] = mso.existing['tenantId']
+        else:
+            mso.existing = mso.request(path, method='POST', data=mso.sent)
```
Comment (shrsr) [inline plugins/modules/mso_dhcp_relay_policy.py]: same as above

File: plugins/modules/mso_dhcp_relay_policy.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+            changed = True
+            response['name'] = payload['name']
+            response['tenant'] = payload['tenantId']
+
+    mso.exit_json(changed=changed, meta=response)
```
Comment (shrsr) [inline plugins/modules/mso_dhcp_relay_policy.py:135]: Can be changed to mso.exit_json()
if the above changes are made

File: plugins/modules/mso_dhcp_relay_policy_provider.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+  state:
+    description:
+    - State of the DHCP Relay provider
+    type: str
+    required: yes
+author:
+- Jorge Gomez Velasquez
+"""
+
+EXAMPLES = r"""
```
Comment (shrsr) [inline plugins/modules/mso_dhcp_relay_policy_provider.py:71]: Examples are required for reference

File: plugins/modules/mso_dhcp_relay_policy.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        desc=description,
+        policyType="dhcp",
+        policySubtype="relay",
+        tenantId=tenant_id,
+    )
+
+    response = {}
+    changed = False
+
+    if state == "query":
```
Comment (xinyuezhao) [inline plugins/modules/mso_dhcp_relay_policy.py:111]: why we pass querying as we can get "mso.existing = mso.query_objs(path, key="DhcpRelayPolicies")" in line 95

File: plugins/modules/mso_dhcp_relay_policy_provider.py [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+        changed = True
+        response["msg"] = "Provider modified"
+        response["tenant"] = mso.existing["provider"]
+    elif not changed_object:
+        response["msg"] = "Provider already exists / Provider does not exits"
+
+    mso.exit_json(changed=changed, meta=response)
+
+
+def check_new_provider(provider_list, provider_to_add, ext_epg_type):
```
Comment (xinyuezhao) [inline plugins/modules/mso_dhcp_relay_policy_provider.py:213]: maybe we want to put this function into mso.py

File: tests/integration/targets/mso_dhcp_relay_policy/tasks/main.yaml [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+
+- name: Verify dhcp_pol1_cm
+  assert:
+    that:
+    - dhcp_pol1_cm is changed
+    - dhcp_pol1_cm.current.name == 'DHCP_REL_POL_1'
+    - dhcp_pol1_cm.current.desc == 'My Test DHCP Policy 1'
+    #- dhcp_pol1_cm.current.tenant == 'ansible_test'   # TODO: Tenant IDs look-up
+
+#TODO: Difference Normal Mode and Check Mode
```
Comment (xinyuezhao) [inline tests/integration/targets/mso_dhcp_relay_policy/tasks/main.yaml:83]: maybe we want to check and fix the difference between normal & check mode

File: tests/integration/targets/mso_dhcp_relay_policy/tasks/main.yaml [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+    state: present
+  register: dhcp_pol2_nm_again
+
+- name: Verify dhcp_pol2_nm_again
+  assert:
+    that:
+    - dhcp_pol2_nm_again is not changed
+
+# ADD EXISTING DHCP POLICY WITH NO STATE
+- name: Add a new DHCP Relay Policy 3  again (normal mode)
```
Comment (xinyuezhao) [inline tests/integration/targets/mso_dhcp_relay_policy/tasks/main.yaml:120]: maybe name it as "Add a new DHCP Relay Policy 3 (normal mode)"

File: tests/integration/targets/mso_dhcp_relay_policy/tasks/main.yaml [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+- name: Verify dhcp_pol2_query
+  assert:
+    that:
+    - dhcp_pol2_query is not changed
+
+# QUERY A NON-EXISTING DHCP RELAY POLICY
+- name: Query non-existing  DHCP Relay Policy 5  (normal mode)
+  mso_dhcp_relay_policy:
+    <<: *mso_info
+    name: DHCP_REL_POL_5
```
Comment (xinyuezhao) [inline tests/integration/targets/mso_dhcp_relay_policy/tasks/main.yaml:150]: you may want to use 'non_existing_DHCP_REL_POL'

**Review by xinyuezhao — CHANGES_REQUESTED** [review]

**Review by anvitha-jain — COMMENTED** [review]
Next module fails, since schema cannot be deleted without deleting the EPG in the template.
To achieve this, the policy needs to be deleted first followed by deleting the EPG -> Template -> Schema.

**Review by lhercot — APPROVED** [review]

File: plugins/modules/mso_dhcp_option_policy.py [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    host: mso_host
+    username: admin
+    password: SomeSecretPassword
+    dhcp_option_policy: my_test_dhcp_policy
+    description: "My Test DHCP Policy"
+    tenant: ansible_test
+    state: absent
+  delegate_to: localhost
+
+- name: Query  a DHCP Option Policy
```
Comment (anvitha-jain) [inline plugins/modules/mso_dhcp_option_policy.py:72]: There is a extra space

File: plugins/modules/mso_dhcp_option_policy.py [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.cisco.mso.plugins.module_utils.mso import MSOModule, mso_argument_spec
+
+
+def main():
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        dhcp_option_policy=dict(type="str", aliases=['name']),
+        description=dict(type="str"),
+        tenant=dict(type="str"),
```
Comment (xinyuezhao) [inline plugins/modules/mso_dhcp_option_policy.py:102]: Is the tenant required when creating DHCP policy?

File: plugins/modules/mso_dhcp_option_policy_option.py [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+DOCUMENTATION = r"""
+---
+module: mso_dhcp_option_policy_option
+short_description: Manage DHCP options in a DHCP Option policy.
+description:
+- Manage DHCP options in a DHCP Option policy on Cisco Multi-Site Orchestrator.
+author:
+- Jorge Gomez (@jorgegome2307)
```
Comment (xinyuezhao) [inline plugins/modules/mso_dhcp_option_policy_option.py:24]: maybe add ''-  Lionel Hercot '' as well

**Review by xinyuezhao — CHANGES_REQUESTED** [review]

File: plugins/modules/mso_dhcp_option_policy.py (hunk 18 above)
Comment (lhercot) [reply plugins/modules/mso_dhcp_option_policy.py:102]: For present yes, not for absent nor query.

**Review by xinyuezhao — APPROVED** [review]

**Review by xinyuezhao — COMMENTED** [review]
LGTM

**Review by anvitha-jain — APPROVED** [review]

**Pull request merged**